Public/Src/Sandbox/Windows/DetoursServices/PolicyResult.cpp

```diff
@@ -102,3 +103,51 @@ void PolicyResult::ReportIndeterminatePolicyAndSetLastError(FileOperationContext
         ERROR_SUCCESS,
         -1);
 }
+
+bool PolicyResult::AllowWrite() const {
+
+    bool isWriteAllowedByPolicy = (m_policy & FileAccessPolicy_AllowWrite) != 0;
+
+    // Send a special message to managed code if the policy to override allowed writes based on file existence is set
+    // and the write is allowed by policy (for the latter, if the write is denied, there is nothing to override)
+    if (OverrideAllowWriteForExistingFiles() && isWriteAllowedByPolicy) {
+        
+        // Let's check if this path was already checked for allow writes in this process. Observe this structure lifespan is the same 
+        // as the current process so other child processes won't share it. 
+        // But for the current process it will avoid probing the file system over and over for the same path.
+        FilesCheckedForAccess* filesCheckedForWriteAccess = GetGlobalFilesCheckedForAccesses();
+        
+        if (filesCheckedForWriteAccess->TryRegisterPath(m_canonicalizedPath)) {
+            DWORD error = GetLastError();

```

Comment:
>DWORD error = GetLastError(); [](start = 12, length = 29)

Why do you need to get the last error here? I think it's just simply ERROR_SUCCESS

Reply:
What I'm after is so last error is preserved. The previous implementation of AllowWrite didn't do any OS calls, so I was not sure if any caller was assuming that AllowWrite didn't change last error. So just making sure ExistsAsFile is not changing that value.

---
In reply to: [299198238](https://github.com/microsoft/BuildXL/pull/536#discussion_r299198238) [](ancestors = 299198238)